Fix ansible-galaxy traceback when unexpected version of resolvelib is installed

s-hertel opened the pull request:
##### SUMMARY
Related to #77624

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

s-hertel committed:
Fix traceback when unexpected version of resolvelib is installed

sivel commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1580,6 +1586,9 @@ def _resolve_depenency_map(
     """Return the resolved dependency map."""
+    if RESOLVELIB_VERSION < RESOLVELIB_LOWERBOUND or RESOLVELIB_VERSION >= RESOLVELIB_UPPERBOUND:
+        raise AnsibleError(f"ansible-galaxy requires resolvelib >= {RESOLVELIB_LOWERBOUND}, < {RESOLVELIB_UPPERBOUND}")
+
```
What you have here is potentially less complicated, but here is an example of doing this without the need of hardcoding versions:

```python
from importlib.metadata import distribution
from packaging.requirements import Requirement

dist = distribution('ansible-core')
req, = [rr for r in dist.requires if (rr := Requirement(r)).name == 'resolvelib']
if not req.contains(resolvelib.__version__):
    raise AnsibleError(f"ansible-galaxy requires {req.name}{req.specifier}")
```

bcoca replied:
just hardcoding it in 'one' place, the requirements file

s-hertel replied:
Oh, thanks for the pointer!

s-hertel committed:
Remove hardcoded resolvelib versions

Co-authored-by: Matt Martz <matt@sivel.net>

s-hertel committed:
Fix test failures

s-hertel commented:
bot_status

s-hertel committed:
Fix running resolvelib built from source

Reorganize to share code with 77649

s-hertel committed:
changelog

s-hertel committed:
Use best-effort to validate the resolvelib version and give a warning if the distribution can't be found

s-hertel committed:
Add a test

s-hertel commented on test/lib/ansible_test/_internal/ansible_util.py:
```diff
@@ -214,35 +211,27 @@ def get_ansible_python_path(args):  # type: (CommonConfig) -> str
-        return
-
-    write_text_file(pkg_info_path, pkg_info.lstrip(), create_directories=True)
+def generate_dist_info():  # type: () -> None
```
@mattclay Does this change look okay to backport? I wanted to add a test and the distribution has no requirements with only the egg-info stub.

mattclay replied:
Why is this change needed?

sivel replied:
The previous stub didn't fully replicate what an egg_info or dist_info would include. So uses of `importlib.metadata` were failing where we needed to query information that was not included in the stub. I recommended generating an actual output, otherwise we'd always be playing catch up to keep it properly in sync with expectations.

mattclay replied:
Unfortunately `setup.py` isn't available once ansible is installed, so we'll need a different solution.

sivel replied:
For that case could we just use the dist_info already generated by setuptools/pip?

Generate it when needed, and copy when available? I don't know the full flow of `ansible-test` for this area of the code to know if that's feasible.

mattclay replied:
We might be able to figure something out, but that doesn't seem like a change we'd want to include in a backport.

What is missing from the current stub?

s-hertel replied:
@mattclay The python requirements are missing. This PR is trying to check the resolvelib version to determine if ansible-galaxy can use it, to preemptively avoid the traceback. I could possible add exception handling instead (TypeError would be the obvious one with the changing function signatures in resolvelib), but making sure all of the unsupported versions would be covered by exception handling doesn't seem practical.

So, I either need to hardcode the supported versions of resolvelib (original attempt, would require code to be updated whenever the requirements are updated), or get the current requirements from the dist-info (current implementation, no upkeep required). 

This patch falls back to the current devel behavior (try to resolve collection dependencies regardless of resolvelib version) if the ansible-core distribution info is wrong (either incomplete with no python requirements, or missing altogether if no 'ansible-core' dist is found). With only the current stub, importlib.metadata.distribution('ansible-core').requires is None, so the ansible-galaxy-collection integration test can not use the bugfix at all.

I didn't want to, but maybe it would be fine to backport without tests?

mattclay replied:
I think a hard-coded min and max version for resolvelib makes sense. Multiple files already need to be updated any time the supported version range changes -- adding one more doesn't seem like a big deal.

s-hertel replied:
I'd rather not add more of it just because it's done elsewhere and is convenient though. I'd like to remove the need to update the test suite but couldn't think of a good way to do so.

And the test doesn't *need* to be updated to function, just should be to validate the range of supported versions.

s-hertel replied:
@mattclay I removed this and added back the hardcoded versions for 2.13 as a fallback (with the hope that I can remove that in devel). How does this look now?

s-hertel committed:
Add a hardcoded fallback for the supported lowerbound/upperbound to make this backportable with tests

Add TODO to update ansible-test's dist info for ansible-core and convert to a warning

mattclay commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1566,6 +1569,23 @@ def _resolve_depenency_map(
+        if req is None:
+            # TODO: replace the hardcoded versions with a warning if the dist info is missing
+            # display.warning("Unable to find 'ansible-core' distribution requirements to verify the resolvelib version is supported.")
+            if RESOLVELIB_VERSION < RESOLVELIB_LOWERBOUND or RESOLVELIB_VERSION >= RESOLVELIB_UPPERBOUND:
```
It's purely cosmetic, but this could be simplified:

```suggestion
            if not RESOLVELIB_LOWERBOUND <= RESOLVELIB_VERSION < RESOLVELIB_UPPERBOUND:
```

mattclay commented on test/integration/targets/ansible-galaxy-collection/tasks/unsupported_resolvelib.yml:
```diff
@@ -0,0 +1,43 @@
+  shell: "{{ ansible_python_interpreter }} -m pip freeze | grep 'resolvelib'"
+  register: installed_resolvelib
+
+- block:
```
I think this test should be moved to its own target and use a runme.sh to setup a venv. Otherwise we risk overwriting an OS package install of resolvelib with pip.

s-hertel replied:
Good catch, fixed. Now the unsupported resolvelib versions are tested using a venv. I also fixed import errors for very outdated (or missing) resolvelib.

s-hertel committed:
Update lib/ansible/galaxy/collection/__init__.py

Co-authored-by: Matt Clay <matt@mystile.com>

s-hertel committed:
Fix tests, use a venv to test handling an unsupported resolvelib versions

s-hertel committed:
Also fix traceback for failing to import resolvelib at all

s-hertel committed:
Fix test?

mattclay committed:
Resolve remaining import sanity test issues.

mattclay commented on test/integration/targets/ansible-galaxy-collection/tasks/main.yml:
```diff
@@ -50,6 +50,12 @@
     src: ansible.cfg.j2
     dest: '{{ galaxy_dir }}/ansible.cfg'
 
+- name: test install command an unsupported version of resolvelib
```
```suggestion
- name: test install command using an unsupported version of resolvelib
```

s-hertel committed:
Update test/integration/targets/ansible-galaxy-collection/tasks/main.yml

Co-authored-by: Matt Clay <matt@mystile.com>

felixfontein commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1566,6 +1588,27 @@ def _resolve_depenency_map(
+                raise AnsibleError(
+                    f"ansible-galaxy requires resolvelib<{RESOLVELIB_UPPERBOUND.vstring},>={RESOLVELIB_LOWERBOUND.vstring}"
+                )
+        elif req.specifier.contains(RESOLVELIB_VERSION.vstring):
```
This should have a `not` in here :)

felixfontein replied:
Fix: #77906

s-hertel replied:
I didn't test manually again after removing the CI coverage. :cold_sweat:
